src/concrete_convex_hull.jl

```diff
@@ -186,6 +191,126 @@ function _three_points_2d!(points::AbstractVector{<:AbstractVector{N}}) where {N
+    A, B, C, D = points[1], points[2], points[3], points[4]
+    tri_ABC = (A[2] - B[2]) * C[1] + (B[1] - A[1]) * C[2] + (A[1] * B[2] - B[1] * A[2])
+    tri_ABD = (A[2] - B[2]) * D[1] + (B[1] - A[1]) * D[2] + (A[1] * B[2] - B[1] * A[2])
+    tri_BCD = (B[2] - C[2]) * D[1] + (C[1] - B[1]) * D[2] + (B[1] * C[2] - C[1] * B[2])
+    tri_CAD = (C[2] - A[2]) * D[1] + (A[1] - C[1]) * D[2] + (C[1] * A[2] - A[1] * C[2])
+    key = 0
+    if tri_ABC > zero(N)
+        key = key + 1000
+    end
+    if tri_ABD > zero(N)
+        key = key + 100
+    end
+    if tri_BCD > zero(N)
+        key = key + 10
+    end
+    if tri_CAD > zero(N)
+        key = key + 1
+    end
+    
+    function collinear_case(A, B, C, D)
+        # A, B and C collinear, D is the extra point
+        if isapprox(A[1], B[1]) && isapprox(B[1], C[1]) && isapprox(C[1], A[1])
+            # points are approximately equal in their first component
+            if isapprox(A[2], B[2]) && isapprox(B[2], C[2]) && isapprox(C[2], A[2])
+                # the three points are approximately equal
+                points[1], points[2] = A, D
+                pop!(points)
+                pop!(points)
+                return _two_points_2d!(points)
+            else
+                # assign the points with max and min value in their second component to the
+                # firsts points and the extra point to the third place, then pop the point that was in the middle
+                points[1], points[2], points[3] = points[argmin([A[2], B[2], C[2]])], points[argmax([A[2], B[2], C[2]])], D
+                pop!(points)
+            end
+        else
+            # assign the points with max and min value in their first component to the
+            # firsts points and the extra point to the third place, then pop the point that was in the middle
+            points[1], points[2], points[3] = points[argmin([A[1], B[1], C[1]])], points[argmax([A[1], B[1], C[1]])], D
+            pop!(points)
+        end
+        return _three_points_2d!(points)
+    end
+    
+    if isapproxzero(tri_ABC)
+        return collinear_case(A, B, C, D)
+    end
+    if isapproxzero(tri_ABD)
+        return collinear_case(A, B, D, C)
+    end
+    if isapproxzero(tri_BCD)
+        return collinear_case(B, C, D, A)
+    end
+    if isapproxzero(tri_CAD)
+        return collinear_case(C, A, D, B)
+    end
+    
+    # ABC  ABD  BCD  CAD  hull
+    # ------------------------
+    #  +    +    +    +   ABC
+    #  +    +    +    -   ABCD
+    #  +    +    -    +   ABDC
+    #  +    +    -    -   ABD
+    #  +    -    +    +   ADBC
+    #  +    -    +    -   BCD
+    #  +    -    -    +   CAD
+    #  +    -    -    -   [should not happen]
+    #  -    +    +    +   [should not happen]
+    #  -    +    +    -   ACD
+    #  -    +    -    +   DCB
+    #  -    +    -    -   DBCA
+    #  -    -    +    +   ADB
+    #  -    -    +    -   ACDB
+    #  -    -    -    +   ADCB
+    #  -    -    -    -   ACB
+    if key == 1111
```

Comment:
A slightly faster implementation (with a bit more to write) is the following:
```julia
# +    +    +    +   ABC
function _four_points_2d_helper!(points, key::Val{1111}, A, B, C, D)
    points[1], points[2], points[3] = A, B, C
    pop!(points)
end
# +    +    +    -   ABCD
function _four_points_2d_helper!(points, key::Val{1110}, A, B, C, D)
     points[1], points[2], points[3], points[4] = A, B, C, D
end
...
# fallback
function _four_points_2d_helper!(points, key, A, B, C, D)
    @assert false "unexpected case in convex_hull"
end

_four_points_2d_helper!(points, key, A, B, C, D)
```
Rationale: The compiler uses a jump table to choose the correct method instead of running along this long if-then-else chain.

These helper methods can also be added inside the function `_four_points_2d!` which may allow to omit the arguments apart from `key` (not sure if this has better, worse, or the same performance).

Reply:
After benchmarking, this turned out to be slower.